OpenCL_C.txt

```diff
@@ -2335,47 +2229,394 @@ defined as the value of the <<opencl-device-queries,
+    }
+}
+----------
+
+Program scope variables or variables with `extern`/`static` storage class
+specifier:
+
+  * Must be qualified by `__constant` in OpenCL C prior to 2.0 or OpenCL C 3.0
+    without `+__opencl_c_program_scope_global_variables+` feature.
+  * Can be qualified by either `__constant` or `__global` for OpenCL C 2.0 or
+    OpenCL C 3.0 with `+__opencl_c_program_scope_global_variables+` feature.
+
+Examples:
+
+[source,c]
+----------
+// Note: these examples assume OpenCL C 2.0 or the
+// __opencl_c_program_scope_global_variables feature macro.
+constant int foo;       // OK.
+global int baz;         // OK.
+global uchar buf[512];  // OK.
+
+static global int bat;  // OK. Internal linkage.
+
+extern constant int foo;  // OK.
+
+
+void func(...)
+{
+    
+  constant static int foo = 1; // OK.
+  global extern int foo;       // OK.
+}
+
+global int * global ptr;          // OK.
+constant int *global ptr=&baz;    // Error, baz is in the global address
+                                  // space.
+global int * constant ptr = &baz; // OK.
+----------
+
+
+Kernel function arguments declared to be a pointer or an array of a type
+must point to one of the named address spaces `+__global+`, `+__local+` or
+`+__constant+`.
+
+Examples:
+
+[source,c]
+----------
+kernel void my_kernel(global int *ptr) // OK
+{
+  ...
+}
+kernel void my_kernel(int *ptr) // Error, ptr must point to either global, local or constant int
+{
+  ...
+}
+----------
+
+--
+
+=== Initialization
+--
+Program scope and `static` variables in the `global` address space are zero
+initialized by default. A constant expression may be given as an initializer.
+
+Variables allocated in the `+__local+` address space inside a kernel function
+cannot be initialized.
+
+Constant address space  variables are required to be initialized and the values
+used to initialize these variables must be a compile time constant.
+
+Private address space objects are not initialized by default, any initializer is
+allowed to be given.
+
+Examples:
+
+[source,c]
+----------
+global int a = 12;      // Initialization is allowed.
+global int b;           // Zero initialized.
+constant int c = 12;    // Initializer is a compile time constant.
+constant int d;         // Error. No initializer provided.
+kernel void my_func(...)
+{
+    local float e = 1;  // Error. Initializer is not allowed.
+
+    local float f;
+    f = 1;              // Allowed
+    private int g;      // Uninitialized.
+    constant int h = g; // Error. Initializer is not a constant expression.
+}
+----------
+
+--
+
+[[addr-spaces-inference]]
+=== Inference
+
+--
+
+Address space qualifiers are not required in many cases. If they are not
+specified explicitly the default address space will be inferred depending
+on the declaration scope and the object type.
+
+There is no syntax to provide address space in the source for some situation,
+therefore only default address space is applicable.
```

Comment:
String literal or temporaries. I mention string literals below but temporaries are a bit harder because C doesn't define that concept well. I am thinking perhaps we should commit this sentence since it is not very clear and not easy to explain either?